### ISSUE TITLE
rewriting of DMRG with Sweep engines

### ISSUE THREAD

**Issue by jhauschild** [tenpy/tenpy#39]
@Lschoon  is working on a rewrite of DMRG. Discussions here

**Comment by Lschoon** [comment]
Note that the rewrite is taking place in [my personal branch](https://github.com/Lschoon/tenpy/branches), rather than in the tenpy repository.

**Comment by Lschoon** [comment]
Update:
- The file [`mps_sweeps.py`](https://github.com/Lschoon/tenpy/blob/engine_rebuild/tenpy/algorithms/mps_sweeps.py) is now essentially finished, modulo some future generalizations and/or inclusion of Mixers in the `Sweep` class.
- I built an implementation of a two-site DMRG engine using the `Sweep` class in [`dmrg.py`](https://github.com/Lschoon/tenpy/blob/engine_rebuild/tenpy/algorithms/dmrg.py). Note that this file is in absolute disarray at the moment, please ignore all but the `TwoSiteDMRGEngine`.

Next order of business is to include (a) mixer(s) in the two-site DMRG algorithm as a proof of concept, and hopefully do some testing.

Any feedback on the work done so far would be very helpful.

**Comment by jhauschild** [comment]
So far it looks very good and I have nothing to complain about :+1: 
You wrote some questions in the ToDo's, I would suggest to discuss in person as long, that's easier.

**Comment by Lschoon** [comment]
Let us discuss those points today, then (if possible).

Two-site DMRG now works completely; I have rewritten [`d_dmrg.py`](https://github.com/Lschoon/tenpy/blob/engine_rebuild/examples/d_dmrg.py) by ways of a test, which now gives the same result as the original did.

Next up:
- The documentation for the new setup still needs some work and cleanup.
- Implement single-site DMRG algorithm.

**Comment by Lschoon** [comment]
Update:
- Single-site DMRG with `combine=False` runs without error. It does not reproduce the correct results yet, as it has no way to grow the bond dimension.
- Single-site DMRG with `combine=True` currently fails at [this line](https://github.com/Lschoon/tenpy/blob/engine_rebuild/tenpy/algorithms/dmrg.py#L699), as the pipes are incorrect. I have yet to figure out how to replace `pipesR` with a trivial pipe in such a way that things don't break. @jhauschild, perhaps you could have a brief look?
- The single-site mixer is only defined for engines with `combine=True`.

To Do:
- [ ] Documentation
- [ ] Make single-site mixer work with `combine=False`
- [ ] Make single-site DMRG work with `combine=True`

**Comment by jhauschild** [comment]
I think in the line you referenced there should only be one pipe, not two?
I think in the single-site DMRG, the combine=True should combine the 'p' leg to the left ('vL') if we move right, or to the right ('vR') if we move left. Thus this whole function 'prepare_update' depends on the sweep movement.

**Comment by Lschoon** [comment]
I agree. Turns out the problem was a small syntax issue leading to `combine_legs` demanding I specify two pipes when I only had one. I believe this is fixed now.

To figure out which direction the sweep is going, it turns out to be not sufficient to look at the `update_LP` or `update_RP` flags (they are `True` too often). I am working around this by setting a global flag `move_right` to `True` by default, switching it to `False` when `i0 == max(schedule_i0)`. Would you agree or do you think there is a cleaner way?

**Comment by Lschoon** [comment]
Single-site DMRG without mixer now runs without errors (but again, without increasing the bond dimension, the results are nonsense).

The `SingleSiteMixer` is giving me a slight headache, not in the least because `subspace_expand()` (which I believe to be bugged) uses numerically indexed legs rather than labels. I am working through the 2015 Hubig paper, but if you could send me a quick sketch of what the idea is, that would speed things up considerably.

**Comment by Lschoon** [comment]
After discussing with @jhauschild, it turns out I made a mistake in `prepare_svd()` by already contracting theta with the next MPS tensor. This is now fixed. Some issues still remain:
- [ ] Running single-site DMRG with `combine=True` and `mixer=True` fails.
- [ ] Running single-site DMRG with any other configuration of those two parameters runs without errors, but gives the wrong result, _even when the mixer is activated_.

I believe both these issues might be due to the indices `update_LP()` and `update_RP()` being off, but haven't fully figured out what's going on yet.

**Comment by jhauschild** [comment]
should I take a quick look?

**Comment by Lschoon** [comment]
I think that would be helpful. My github repo is up-to-date.

One thing also worth thinking about is if the schedules for 1- and 2-site DMRG are the same for finite DMRG. I think the rightmost site might not be updated right now, but changing the schedule seems to lead to an indexing error in the MPS.

**Comment by Lschoon** [comment]
About this last issue. I am now convinced that the schedule for single-site algorithms was wrong and missed the rightmost site. I fixed this in [a recent commit](https://github.com/Lschoon/tenpy/commit/dddc4c154422df9e2d861550ffba12628f078df0). This now leads to some issues where `get_LP` or `get_RP` try to get out-of-bounds MPS tensors.

**Comment by jhauschild** [comment]
Agreed. I'm already working on it and changed this locally. I will soon upload my commit.

**Comment by Lschoon** [comment]
Alright, I will hold off on more updates until your commit is in.

**Comment by Lschoon** [comment]
With @jhauschild's help, single-site DMRG for finite systems now works correctly. Infinite systems not yet.

- [ ] Debug 1-site DMRG for infinite systems
- [ ] Documentation

**Comment by Lschoon** [comment]
As per https://github.com/Lschoon/tenpy/commit/14c877b90bfdca21c3f8b1845be7f271f1daf825, single-site DMRG works as expected in all test cases! The bugfix for the infinite system turned out to be an indexing error in `update_RP()`.

I will now focus on writing documentation.

**Comment by jhauschild** [comment]
That's great news! One additional thing:
- [ ] make sure that test/test_dmrg.py doesn't fail (using `pytest`). There might still be some bugs in orthogonal_to, for example.

**Comment by Lschoon** [comment]
Good point! I have changed test_dmrg.py to work with the new engines. Two issues remain here:
- 1-site infinite DMRG does not reach the required threshold in energy (the difference with the exact energy is around `1e-8`, while tests fail for errors bigger than `1e-10`). I tried playing around with the mixer parameters, but this problem persists.
- (probably unrelated) `test_model_toric_code.py` fails a test as the ground state energy is way off. Here,  even 2-site DMRG does not change the MPS bond dimension, so the state remains a product state. No idea why this happens...

**Comment by Lschoon** [comment]
All class and method docstrings are finished and ready for review. Also, I changed the global `dmrg.run()` function to work with the new engines.

Remaining tasks:
- [ ] Global docstring for `mps_sweeps.py`
- [ ] The `TwoSiteMixer` has not been ported yet, this could potentially be interesting (but not absolutely necessary).
- [ ] Fix precision issue with 1-site iDMRG (see above).
- [ ] Fix toric code model problem (see above).
- [ ] (Medium term) Move all methods that `OneSiteDMRGEngine` inherits from `TwoSiteDMRGEngine` to a `DMRGEngine` superclass.
- [ ] (Medium term) Port plotter functions from old `Engine` class.
- [ ] (Medium term) Include orthogonal theta's in `EffectiveH`'s. Not sure how/why anymore.
- [ ] (Long term, should probably be done by someone who understands TDVP) Port TDVP to the new `Sweep`-based setup.

**Comment by Lschoon** [comment]
@jhauschild, I could use your input with the issue I mentioned above. I initially thought this was some problem with the DMRG parameters in `test_dmrg.py`, but I have played with those a bit without affecting the problem at all. The issue is as follows:
- 1-site DMRG with `bc='infinite'` does not reach the threshold set by `test_dmrg.py`.
- (Reported) convergence is good, reaching `norm_err = 1.7e-13, Delta E = -1.4033e-12, Delta S = -7.7925e-11` on my last test before hitting `max_sweeps`.
- `chi_max` is not saturated and, perhaps crucially, the final bond dimension is lower than in the 2-site case. However, when I decrease `svd_min`, the bond dimension grows but the final result gets _worse_.

I am not sure where to start debugging this, any suggestions would be helpful.

**Comment by jhauschild** [comment]
I looked a little bit into the problem and observed the following:

- While the energy returned by DMRG is off, the energy calculated with `np.mean(M.bond_energies(psi))` seems to be correct down to Machine precision for `g=1.5` in the transverse field Ising model.
I think this is due to the mixer screwing up the orthonormality of the left/right states in the infinite case.
- The default 15 sweeps with activated mixer will often be too little for 1-site DMRG.
- Too small mixer amplitude compared to `svd_min` is bad, because it doesn't allow to grow the bond dimension with the mixer. Too large mixer amplitude on the other hand screws up the state (in particulare the `norm_err`) quite badly. I'm not sure what the best values are...
- The toric code test works, if one puts `Lx=2` instead of `Lx=1`.
  Reason: all of the star terms getting us out of the initial product state extend over more than one 'ring'/MPS unit cell.
  In the old Code, we kept 'mixing' with the left part of the effective Hamiltonian whenever we updated the left part, while the new code we only mix on the left while moving right. We might need to change this.

**Comment by Lschoon** [comment]
Thank you for looking into this! I took the liberty of numbering your comments for easier discussion.

1. Indeed; in fact every single bond energy is has a reasonable error. I don't fully understand your comment about the orthonormality of left/right states.
2. True, but this is not why the tests fail (as you probably found yourself).
3. As in 2. These probably requires some amount of tinkering depending on the model.
4. That fixed it indeed! I am still curious why this did not cause an issue before the rewrite.
5. Again, not completely sure what you mean here. Do you mean the old (Python 2) tenpy, or the current master build?

**Comment by jhauschild** [comment]
1. In the finite case, we need the canonical form only to ensure the orthonormality `A^dagger A = eye` and `B B^dagger = eye` of the left and right schmidt states. This is ensured by the SVD we do after the mixing. However, the mixing screws up the normalization: if we move right, we do:
```python
theta, next_B = subspace_expand(theta, next_B)  # add's zeros in next_B
U, S, VH = svd(theta) 
# I think, here norm(U S)=1 and norm(S VH)=1, but
#  norm(dot(S VH, next_B)) != 1 because of the extra zeros in next_B.
next_B = dot(VH, next_B)
```
In the infinite case, we absorb the next_B in the right environment when changing direction, which then screws up the normalization of the envirionments. In the end, I think it's just a global prefactor in the environment tensors, which screws up the energy values, but still leads to the correct ground state - if you calculate the overlap with the ground state from two-site DMRG without mixer, it's 1.
2. Sure, just wanted to mention it
3. As 2.
4. This is still a severe issue, changing the test doesn't fix the problem. I think the reason is what I tried to explain in 5. Let me try again:
5. Here, old=both "before the rewrite" and "old Tenpy", they behave the same way.
  Before the rewrite, we decided with `update_LP, update_RP` whether to mix on the left or rigtht. Considering the schedule, this leads to *more* mixing, in particular to continuously mix on the left whenever we grow the left environment. In contrast, after the rewrite, we stop mixing on the left the moment we switch to a right move, even though we might update the left part of the environment.
It might be necessary to do this to properly account for long-range terms.

**Comment by Lschoon** [comment]
This comment is mostly to track my own thoughts while debugging; however, any input is still very welcome. I am in a world of confusion about normalizations...

- Regarding the normalization of the left/right environments: these are non-normalized in the infinite case even when no mixer is used. That does not mean there is no problem with an overall factor in the environments. There seems to be a difference between finite/infinite here: in the finite case, the leftmost/rightmost environments _are_ normalized to 1. The leftmost env starts out 1-normalized in the infinite case as well, but changes during the sweeps. I am not convinced this is the expected behaviour
- In the finite case, the normalization of the MPS tensors is also not equal to 1 except at the edge of the system. This is independent of mixer use. 
- @jhauschild's comment in the code above: 
```
# I think, here norm(U S)=1 and norm(S VH)=1, but
#  norm(dot(S VH, next_B)) != 1 because of the extra zeros in next_B.
```
is indeed correct.

Given all this, and the discussion above about the scheduling of updates to LP/RP, I am not fully convinced there is a problem with normalizations. Rather, I have a suspicion that somewhere, an environment is wrongly updated (or not updated when it should be) due to a scheduling or related error.

**Comment by Lschoon** [comment]
576120d4926b21646aea62b7ddea9bdcf586dead solves the problem of the DMRG energy not being accurate (enough) by enforcing normalization of the A/B tensors explicitly after (mixed) SVD. This is as we discussed above.

There is some potential for this code to be prettified or otherwise improved, so any feedback would be very welcome. In particular, it would probably be good to only enforce the normalization when the mixer is active, to avoid extra overhead when we don't need it.

**Comment by Lschoon** [comment]
a3134cdd89d1b237ccb5abb79236c74424d895df resolves the issue found in the toric code test, but inadvertently introduces some problems with finite systems.
